File: ts/src/util/bytes.ts
/**
 * Converts a Uint8Array to a BigInt.
 *
 * @param bytes - The byte array to convert.
 *
 * @returns bigint - The resulting BigInt.
 */
function bytesToBigInt(bytes: Uint8Array): bigint {
    let result = 0n;
    for (const byte of bytes) result = (result << 8n) + BigInt(byte);
    return result;
}

/**
 * @note This function assumes a big-endian representation.
 *
 * Converts a BigInt to a Uint8Array.
 *
 * @throws {RangeError} - If the input BigInt is negative.
 *
 * @param value - The BigInt to convert.
 *
 * @returns Uint8Array - The resulting byte array.
 */
function bigIntToByteArray(value: bigint): Uint8Array {
    if (value < 0n) {
        throw new RangeError("bigIntToByteArray only supports non-negative values");
    }

    if (value === 0n) {
        return new Uint8Array([0]);
    }

    const bytes: number[] = [];
    let v = value;

    while (v > 0n) {
        bytes.push(Number(v & 0xffn));
        v >>= 8n;
    }

    bytes.reverse();

    return new Uint8Array(bytes);
}

/**
 * @note This function returns the big-endian representation.
 * @important For lengths greater than 53 bits this function may not work as expected due to JavaScript's number limitations.
 *
 * Converts an integer to a Uint8Array of specified byte length.
 *
 * @throws {RangeError} - If the input integer is negative or exceeds the safe integer range.
 *
 * @param i - The integer to convert.
 * @param byteLength - The desired byte length of the output array.
 *
 * @returns Uint8Array - The resulting byte array.
 */
function intToBytes(i: number, byteLength: number): Uint8Array {
    if (i < 0) throw new RangeError("intToBytes only supports non-negative integers");
    if (!Number.isSafeInteger(i)) throw new RangeError("Input exceeds 53-bit safe integer range");

    const bytes = new Uint8Array(byteLength);
    let value = BigInt(i);

    for (let b = byteLength - 1; b >= 0; b--) {
        bytes[b] = Number(value & 0xffn);
        value >>= 8n;
    }

    return bytes;
}

/**
 * Concatenates multiple Uint8Array instances into a single Uint8Array.
 *
 * @param arrays - The arrays to concatenate.
 *
 * @returns Uint8Array - The concatenated byte array.
 */
function concatBytes(...arrays: Uint8Array[]): Uint8Array {
    const totalLength = arrays.reduce((sum, arr) => sum + arr.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;

    for (const arr of arrays) {
        result.set(arr, offset);
        offset += arr.length;
    }

    return result;
}

/**
 * @note This function assumes a big-endian representation.
 *
 * Creates a framed byte array with a length prefix.
 *
 * @param data - The data to frame.
 * @param lengthPrefixBytes - The number of bytes to use for the length prefix.
 *
 * @returns Uint8Array - The framed byte array.
 */
function framedBytesFromUint8Array(data: Uint8Array, lengthPrefixBytes: number): Uint8Array {
    return concatBytes(intToBytes(data.length, lengthPrefixBytes), data);
}

/**
 * @note This function assumes a big-endian representation.
 *
 * Creates a framed byte array from a BigInt with a length prefix. This framing includes a sign byte which reflects
 * the GoLang implementation.
 *
 * @param value - The BigInt value to frame.
 * @param lengthPrefixBytes - The number of bytes to use for the length prefix.
 *
 * @returns Uint8Array - The framed byte array.
 */
function framedBytesFromBigInt(value: bigint, lengthPrefixBytes: number): Uint8Array {
    const isNegative = value < 0n ? 1 : 0;
    const magnitude = bigIntToByteArray(isNegative ? -value : value);
    return concatBytes(
        intToBytes(magnitude.length, lengthPrefixBytes),
        new Uint8Array([isNegative]),
        magnitude
    );
}

/**
 * @note This function assumes a big-endian representation.
 *
 * Creates a framed byte array from a string with a length prefix.
 *
 * @param str - The string to frame.
 * @param lengthPrefixBytes - The number of bytes to use for the length prefix.
 *
 * @returns Uint8Array - The framed byte array.
 */
function framedBytesFromString(str: string, lengthPrefixBytes: number): Uint8Array {
    const bytes = new TextEncoder().encode(str);
    return concatBytes(intToBytes(bytes.length, lengthPrefixBytes), bytes);
}

/**
 * @note This function assumes a big-endian representation.
 *
 * Generic function to create a framed byte array from various input types with a length prefix.
 * The default length prefix size is 4 bytes.
 *
 * @param input - The input data to frame (Uint8Array, bigint, or string).
 * @param lengthPrefixBytes - The number of bytes to use for the length prefix (default is 4).
 *
 * @returns Uint8Array - The framed byte array.
 */
function framedBytes(input: Uint8Array | bigint | string, lengthPrefixBytes: number = 4): Uint8Array {
    if (input instanceof Uint8Array)  return framedBytesFromUint8Array(input, lengthPrefixBytes);
    else if (typeof input === "bigint")  return framedBytesFromBigInt(input, lengthPrefixBytes);
    else return framedBytesFromString(input, lengthPrefixBytes);
}

export {
    bytesToBigInt,
    bigIntToByteArray,
    intToBytes,
    concatBytes,
    framedBytesFromUint8Array,
    framedBytesFromBigInt,
    framedBytesFromString,
    framedBytes
};

File: go/util/bytes.go
package util

import (
	"errors"
	"math/big"
)

// BytesToBigInt converts a slice of bytes into a big.Int (big-endian).
func BytesToBigInt(b []byte) *big.Int {
	bi := new(big.Int)
	bi.SetBytes(b)
	return bi
}

// BigIntToByteArray converts a non-negative *big.Int to a big-endian byte slice.
// Returns []byte{0} for zero. Returns an error for negative or nil input.
func BigIntToByteArray(v *big.Int) ([]byte, error) {
	if v == nil {
		return nil, errors.New("BigIntToByteArray: input is nil")
	}
	if v.Sign() < 0 {
		return nil, errors.New("BigIntToByteArray only supports non-negative values")
	}
	if v.Sign() == 0 {
		return []byte{0}, nil
	}
	return v.Bytes(), nil
}

// IntToBytes converts a non-negative integer to a big-endian byte slice of fixed length.
// Returns an error if the integer is negative or doesn't fit into the specified byte length.
func IntToBytes(i int64, byteLen int) ([]byte, error) {
	if byteLen <= 0 {
		return nil, errors.New("IntToBytes: byteLen must be positive")
	}
	if i < 0 {
		return nil, errors.New("IntToBytes only supports non-negative integers")
	}

	v := big.NewInt(i)
	limit := new(big.Int).Lsh(big.NewInt(1), uint(8*byteLen))
	if v.Cmp(limit) >= 0 {
		return nil, errors.New("IntToBytes: input does not fit in the requested byte length")
	}

	out := make([]byte, byteLen)
	for b := byteLen - 1; b >= 0; b-- {
		out[b] = byte(new(big.Int).And(v, big.NewInt(0xff)).Uint64())
		v.Rsh(v, 8)
	}
	return out, nil
}

// ConcatBytes concatenates multiple byte slices into one.
func ConcatBytes(arrays ...[]byte) []byte {
	total := 0
	for _, a := range arrays {
		total += len(a)
	}
	res := make([]byte, total)
	offset := 0
	for _, a := range arrays {
		copy(res[offset:], a)
		offset += len(a)
	}
	return res
}

// FramedBytesFromUint8Array prefixes the length (big-endian) of data using lengthPrefixBytes.
func FramedBytesFromUint8Array(data []byte, lengthPrefixBytes int) ([]byte, error) {
	prefix, err := IntToBytes(int64(len(data)), lengthPrefixBytes)
	if err != nil {
		return nil, err
	}
	return ConcatBytes(prefix, data), nil
}

// FramedBytesFromBigInt creates a framed byte array from a big.Int.
// Format: [length (lengthPrefixBytes)][sign byte (0 or 1)][magnitude bytes]
func FramedBytesFromBigInt(value *big.Int, lengthPrefixBytes int) ([]byte, error) {
	if value == nil {
		return nil, errors.New("FramedBytesFromBigInt: value is nil")
	}
	isNegative := byte(0)
	v := new(big.Int).Set(value)
	if v.Sign() < 0 {
		isNegative = 1
		v.Abs(v)
	}
	magnitude, err := BigIntToByteArray(v)
	if err != nil {
		return nil, err
	}
	prefix, err := IntToBytes(int64(len(magnitude)), lengthPrefixBytes)
	if err != nil {
		return nil, err
	}
	return ConcatBytes(prefix, []byte{isNegative}, magnitude), nil
}

// FramedBytesFromString creates a framed byte array from a UTF-8 string.
func FramedBytesFromString(s string, lengthPrefixBytes int) ([]byte, error) {
	bytes := []byte(s) // Go strings are UTF-8 encoded
	return FramedBytesFromUint8Array(bytes, lengthPrefixBytes)
}

// FramedBytes is a generic dispatcher. Input can be []byte, *big.Int, or string.
// lengthPrefixBytes specifies number of bytes used for the length prefix.
func FramedBytes(input interface{}, lengthPrefixBytes int) ([]byte, error) {
	switch v := input.(type) {
	case []byte:
		return FramedBytesFromUint8Array(v, lengthPrefixBytes)
	case *big.Int:
		return FramedBytesFromBigInt(v, lengthPrefixBytes)
	case string:
		return FramedBytesFromString(v, lengthPrefixBytes)
	default:
		return nil, errors.New("FramedBytes: unsupported input type")
	}
}
